### DISCUSSION TITLE
Filters for non-uniformly sampled signals

### DISCUSSION THREAD

**Discussion by RobinSchmidt** [RobinSchmidt/RS-MET#327] [Show and tell]
i want to add filters to the library that allow filtering of non-uniformly sampled data. i found a nice paper about this here:

http://inf.ufrgs.br/~eslgastal/NonUniformFiltering/

i have already implemented the non-uniform one-pole filter for the real and complex case. the real one-pole has an exponential decay as impulse reponse and the complex one-pole has an exponentially enveloped sinusoid. here are both impulse responses - the gray continuous function is the underlying analog/continuous-time impulse response and the dots are (non-unformly sampled) impulse-response output-samples of the current implementation:

![image](https://user-images.githubusercontent.com/1833099/59587615-f4c56b00-90e5-11e9-8e1d-f8f245e4730a.png)

![image](https://user-images.githubusercontent.com/1833099/59587622-fa22b580-90e5-11e9-906b-1507995cee70.png)

this looks quite promising already (in the first plot, some dots look slightly wrong, but that seems to be a plotting artifact).

the eventual goal is to have at some stage all the nice filter types from engineers filter also available for the non-uniform case. i have a quite concrete idea, what has to be done for all-pole filters (which includes butterworth, chebychev-1, bessel, papoulis, halpern, gaussian) - but for filters with zeros (chebychev-2, elliptic), i'm not so sure yet (oh - and at the moment, i'm just talking about lowpasses). i don't fully understand certain things in the paper - they don't explain, how the FIR part is supposed to be implemented. in their code, they seem to just treat it like a uniform FIR (if i interpret it correctly - i don't "speak" julia) - but i don't get, why that should work. also, the computation their "ground-truth" impulse responses (in figure 5) seems a bit fishy to me - we'll see....

**Comment by elanhickler** [comment]
What's the difference between this and interpolation algorithms?

**Comment by RobinSchmidt** [comment]
interpolation produces new datapoints at arbitrary positions in between the existing ones. filters (uniform or not) modify the existing datapoints *at* their original positions. look at figure 6 in the paper. do you remember this `nonUniformMovingAverage` function, i once wrote for you? (still in the experiments project - i should drag over the function into the library). this here is similar in the result/goal - but uses a recursive/IIR filtering algorithm.

**Comment by RobinSchmidt** [comment]
to implement this, i will need to work out some interesting and potentially more generally useful new math algorithms. specifically, i will need a partial fraction expansion. i already have a core routine for this task in class `rsPolynomial` - but it's inconvenient to use and doesn't cover all possible cases (edit: or - looking at it-  it seems, it does cover all cases but foregoes an important optimization opportunity for the (very common and in our case occurring) special case of all poles being distinct). i want to create a new class for rational functions to handle these things more properly and conveniently.

**Comment by RobinSchmidt** [comment]
it seems like i have a nonuniform butterworth filter halfway working - well - at least, the impulse resonses look good (these are non-uniform butterworth impulse responses of orders 1-8):
![image](https://user-images.githubusercontent.com/1833099/60901244-e9410c00-a26d-11e9-9d54-b3fce81b09ab.png)

these do indeed look like Butterworth impulse responses are supposed to look - compare, for example, to this (bottom left plot):
https://ars.els-cdn.com/content/image/3-s2.0-B9780750687034000080-f08-15-9780750687034.jpg?_
https://www.sciencedirect.com/topics/engineering/butterworth-filters

...however - the step responses of the very same filters look like total garbage! 
![image](https://user-images.githubusercontent.com/1833099/60901442-4341d180-a26e-11e9-91ca-16bfa0d76c88.png)

it's literally the same set of filters just with a different input :-O ...apparently, i'm still doing something wrong!

**Comment by RobinSchmidt** [comment]
aha! it seems like i need these DC normalization formulas described in the paper (sections 3.2.3 and 3.2.4). going back to the simple one pole (from the 1st post) - 

impulse response without normalization:
![image](https://user-images.githubusercontent.com/1833099/60903822-522a8300-a272-11e9-90fc-d58fc3f54cf0.png)

impulse response with spatially variant scaling:
![image](https://user-images.githubusercontent.com/1833099/60903908-71291500-a272-11e9-8a0d-74d5ed4b4879.png)

impulse response with piecewise resampling:
![image](https://user-images.githubusercontent.com/1833099/60903962-89009900-a272-11e9-9060-4b1fa511eee2.png)

so - it looks like i can get exact matches to the ideal continuous time impulse-response, if i don't do any normalization - but doing the normalization somehow "disturbs" the non-uniform impulse response. ..but...

**Comment by RobinSchmidt** [comment]
for the step responses, the situation seems reversed:

step response without normalization:
![image](https://user-images.githubusercontent.com/1833099/60904225-f8768880-a272-11e9-97ef-3c59e95e3fb7.png)

step response with spatially variant scaling:
![image](https://user-images.githubusercontent.com/1833099/60904282-15ab5700-a273-11e9-9cc7-d8de72f3c9ab.png)

step response with piecewise resampling:
![image](https://user-images.githubusercontent.com/1833099/60904352-35427f80-a273-11e9-9458-79fe4078f84b.png)

here, doing the normalization gives a better fit and not doing it seems to "disturb" the step response. it seems like the "piecewise resampling" method yields an exact match at the sample instants

**Comment by RobinSchmidt** [comment]
...so that means, i must now implement these normalization schemes for the high-order non-uniform filter. the paper says something about not normalizing the individual filters but the whole sum at once - so i can't just use my existing first-order normalization code. ...more work to do but i have some directions now....

**Comment by RobinSchmidt** [comment]
i wonder, if it makes sense to let the user/client-code somehow use intermediate settings between normalizing and not normalizing, i.e. dial in a compromise between "good impulse response" and "good step response"...like applying the normalization schemes partially

**Comment by RobinSchmidt** [comment]
ok - these non-uniform butterworth step responses look much better:
![image](https://user-images.githubusercontent.com/1833099/60982892-4786f100-a339-11e9-9ffb-9b96dcba36da.png)
the 1st order one shoots slightly above 1 but the other ones look good. i'm getting closer...

**Comment by RobinSchmidt** [comment]
here are two other relevant papers which are cited in the gastal/oliveira paper:

http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.533.3670&rep=rep1&type=pdf
http://ljk.imag.fr/membres/Brigitte.Bidegaray/Sources/FB10.pdf

...marked for reading. they seem to take a different approach. edit: yep - totally different algorithm. they express the filter in continuous time as differential equation and implement numerical ODE solvers. i think, the gastal/oliveira approach can be seen as sort of implementing an analytic solution (due to the explicit call to the exponential function - the b^(delta_t) thing) and deriving a recursive algorithm for advancing that solution from one datapoint ot the next

oh - and here is a full book about non-uniform sampling:

https://books.google.de/books?id=givsYJZyf0gC&printsec=frontcover&source=gbs_ge_summary_r&cad=0#v=onepage&q&f=false

**Comment by RobinSchmidt** [comment]
i have implemented some new algorithm for finding the partial fraction epxansion based on this:

https://github.com/RobinSchmidt/RS-MET/blob/master/Notes/PartialFractions.txt

...the former algorithm relied on establishing and solving a linear system of equations and is probably less efficient and less numerically precise (although i did not check that yet). we can now apply the partial fraction expansion also to filters that have repeated poles. ...although, all the filter types used in EngineersFilter do not actually have repeated poles, i wanted to have that for completeness. it might be relevant for linkwitz-riley filters - maybe it could also be useful for the moog ladder
